python/paddle/utils/code_gen/legacy_api.yaml

```diff
@@ -1525,6 +1525,16 @@
     func : multiply
   backward : multiply_grad
 
+- api : nan_to_num
+  args : (Tensor input, float nan, bool replace_posinf_with_max, float posinf, bool replace_neginf_with_min, float neginf)
```

Comment:
> In order to reduce the threshold for users to use C++ API, it is necessary to ensure that the experience of using C++ API is the same with Python

以现在的 paddle 基础设施来说，这个并不是一定能做到的？python 的类型是 nullable 的，但 C++ 不是，而且 paddle yaml codegen 看起来不支持 std::optional\<T\>，所以如果 Python API 参数为 None 是有意义的，那么 C++ API 就没办法和 Python 直接对齐。这是 nan_to_num c++ api 里有两个额外的 bool 的原因。顺便说一下，yaml codegen 在我最开始实现 nan_to_num 的时候连 double 都不支持，不知道现在修复了没有

> or in the simplest way, use paddle.where to replace nan/posinf/neginf in python function without implement C++ API and kernel?

这是最简单也是最合理的，但和 https://github.com/PaddlePaddle/Paddle/issues/40329 的命题作文不符😂，我是看了这里的要求才实现 kernel 的：

> C ++/CUDA 实现代码，头文件在 Paddle repo 的 [paddle/phi/kernels](https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/phi/kernels)目录，cc文件在[paddle/phi/kernels/cpu](https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/phi/kernels/cpu) 目录 和 cu文件在[paddle/phi/kernels/gpu](https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/phi/kernels/gpu) 目录

Reply:
@jeff41404

Reply:
> > or in the simplest way, use paddle.where to replace nan/posinf/neginf in python function without implement C++ API and kernel?
> 
> 这是最简单也是最合理的，但和 “命题作文”（#40329 ）的要求不符😂，我是看了这里的要求才实现 kernel 的：
> 
> > C ++/CUDA 实现代码，头文件在 Paddle repo 的 [paddle/phi/kernels](https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/phi/kernels)目录，cc文件在[paddle/phi/kernels/cpu](https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/phi/kernels/cpu) 目录 和 cu文件在[paddle/phi/kernels/gpu](https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/phi/kernels/gpu) 目录

if we consider it is the simplest and most reasonable way to implement without C++ API and kernel, we can choose right road, I understand the description of task is a general guide not composition with assigned subject.

Reply:
> > In order to reduce the threshold for users to use C++ API, it is necessary to ensure that the experience of using C++ API is the same with Python
> 
> 以现在的 paddle 基础设施来说，这并不是一定能做到的。python 的类型是 nullable 的，但 C++ 不是，而且 paddle yaml codegen 看起来不支持 std::optional<T>，所以如果 Python API 的参数为 None 是有意义的，那么 C++ API 就没办法和 Python 直接对齐。这是 nan_to_num c++ api 里有两个额外的 bool 的原因，也是和 nanmedian 情况不同的地方。顺便说一下，yaml codegen 在我最开始实现 nan_to_num 的时候连 double 都不支持，不知道现在修复了没有

yaml codegen is testing and checking, `optional` and `double ` are really known problem and we are fixing, we hope it will be basically perfect next month.

Reply:
> if we consider it is the simplest and most reasonable way to implement without C++ API and kernel, we can choose right road, I understand the description of task is a general guide not composition with assigned subject.

您说的我完全认同。不过如果我在报名 hackathon 这个任务的时候，就提出这个算子能通过 paddle.where 拼出来，可能是不会被采纳的😂 毕竟 issue 里确实明确写着“提交内容：C++/CUDA 实现代码”，虽然您作为 paddle 资深开发者有脱离这套表面规则自由发挥的底气，但我作为 hackathon 的众多报名者之一，是应该谨慎行事的。

<img width="274" alt="image" src="https://user-images.githubusercontent.com/20833108/176352691-a1d5f6aa-bec8-454c-8cfc-a7ac8b18451c.png">


另一方面，hackathon 现在已经结束了，这个 PR 该怎么处理呢，如果现在要求改成用 paddle.where 来拼，可能是超出了 hackathon 的范围的

Reply:
>不过如果我在报名 hackathon 这个任务的时候，就提出这个算子能通过 paddle.where 拼出来，可能是不会被采纳的😂 毕竟 https://github.com/PaddlePaddle/Paddle/issues/40329 里确实明确写着“提交内容：C++/CUDA 实现代码”，虽然您作为 paddle 资深开发者有打破这套表面规则并自由发挥的底气，但我作为 hackathon 的众多报名者其中之一，在报名阶段是没有这样的光环的😂。也许更应该做的是在发布 hackathon issue 的时候避免这类问题？

Yes, this really reflects our lack of precision in describing the task, which we have improved in the coming hackathon.

>另一方面，hackathon 现在已经结束了，这个 PR 该怎么处理呢，如果现在要求改成用 paddle.where 来拼，可能是超出了 hackathon 的范围的。如果 nan_to_num 的算子功能已经完整，也和当时的 rfc 相符，感觉可以先合并。

I confirmed with relevant internal colleague in recent days, because this PR was approved before deadline, so it will not be affected by the end time of the previous hackathon, so bonuses still vaild after PR merge. Don't worry beyond the limit of hackathon, just follow the right path, I support you:)